feat: hide balance

adrianocola opened the pull request:
### This PR:
* Fixes #5096 .
* Instead of hiding balance when clicking the balance text, an eye icon was added to hide/show the balance.
* The total USD balance, assets balances, and the values of activities can be hidden.
* Created integration test.

Implementation details:
* Created `HideableBalance` component that shows/hides its child based on the `HideBalanceContext` value.
* Some components that show balance are used everywhere in the code, but only the balances on the home screen should be hidden, so the visibility of balances was conditioned to a React Context.
* Created `HideBalanceProvider` and added it to the home page.
* The hide/show status is stored in Redux and thus persists between restarts.

**Please let me know if there’s anything I need to change!**

Wallet:

https://github.com/user-attachments/assets/e3037ce2-358a-4f1b-9d24-77ddddba62c3



Large Screen:

https://github.com/user-attachments/assets/db0b55a6-dc07-4d20-9b72-0322e917eaaa

adrianocola commented on src/app/ui/components/account/account.card.tsx:
```diff
@@ -37,35 +51,46 @@ export function AccountCard({
+            </Box>
+          </Flex>
+        </Link>
+        <Box mr={{ md: 'space.05' }}>
```
Add this to align with the other buttons ("large screen" only)

adrianocola commented on src/app/common/hide-balance-provider.tsx:
Not sure if this file should be in the "common" directory. Is there a better place?

pete-watters commented:
Thanks for the contribution @adrianocola , its much appreciated. 

I'll check it out now. In the meantime can you please:
- rebase your work with the latest from `dev` (rebase instead of merge)
- squash your commits to a single commit 
- run `prettier` to fix lint issues failing in `lint-prettier`

Thanks!

adrianocola commented:
Hey @pete-watters, thank you for the feedback. I just made the changes you asked for. Let me know if there is anything else I need to change!

pete-watters commented:
> Hey @pete-watters, thank you for the feedback. I just made the changes you asked for. Let me know if there is anything else I need to change!

Thanks @adrianocola 👍 I checked with the team and their feedback was that it would be better to store this setting in `redux` only and avoid using an extra context. Could you refactor it a bit to avoid the extra context?

Otherwise the code looks pretty good to me. Thanks again!

adrianocola commented:
Thanks for the team input, @pete-watters! 

I've added the Context to hide the balances that are only visible on the Home screen. The components that show the balance on the Home screen (like `CryptoAssetItemLayout`) are used on many other screens. Without the context, those balances would be hidden everywhere.

I can hide the balance everywhere in the extension. Some approaches:
1. **Hide/Show Balance Only from the Home Screen**
While this approach ensures that the visibility control is limited to the Home screen, it might frustrate users. They would need to navigate back to the Home screen every time they wish to toggle the visibility.
2. **Hide/Show Balance Everywhere**
If we extend the balance visibility toggle across all screens, the “eye” icon should be accessible on every screen. The obvious place would be the header, but it doesn't feel right. The header is also not always visible (full-screen popups hide the header). An alternative could be to display the “eye” icon in the header only when the balances are hidden, allowing users to reveal them from almost any screen.
3. **Toggle Balance Visibility by Clicking the Hidden Balance Dots**
Another option is to allow users to toggle the visibility by clicking on the hidden balance dots (•••) wherever they appear. However, this introduces complexity, especially on components with interactive/clickable elements like activity items. Making the dots clickable (make them stand out on hover) or adding a small “eye” icon next to them (on hover) could be solutions, but this might lead to usability or accessibility issues.

To maintain the current solution where balances are hidden only on the Home screen, I can modify some components, such as `AssetList` and `ActivityList`, to accept an optional `hideBalance` prop. This prop can then be passed down the component tree until it reaches the actual balance component.

Please let me know what you prefer or if you have any other suggestions!

pete-watters commented:
> Thanks for the team input, @pete-watters!
> 
> I've added the Context to hide the balances that are only visible on the Home screen. The components that show the balance on the Home screen (like `CryptoAssetItemLayout`) are used on many other screens. Without the context, those balances would be hidden everywhere.
> 
> I can hide the balance everywhere in the extension. Some approaches:
> 
> 1. **Hide/Show Balance Only from the Home Screen**
>    While this approach ensures that the visibility control is limited to the Home screen, it might frustrate users. They would need to navigate back to the Home screen every time they wish to toggle the visibility.
> 2. **Hide/Show Balance Everywhere**
>    If we extend the balance visibility toggle across all screens, the “eye” icon should be accessible on every screen. The obvious place would be the header, but it doesn't feel right. The header is also not always visible (full-screen popups hide the header). An alternative could be to display the “eye” icon in the header only when the balances are hidden, allowing users to reveal them from almost any screen.
> 3. **Toggle Balance Visibility by Clicking the Hidden Balance Dots**
>    Another option is to allow users to toggle the visibility by clicking on the hidden balance dots (•••) wherever they appear. However, this introduces complexity, especially on components with interactive/clickable elements like activity items. Making the dots clickable (make them stand out on hover) or adding a small “eye” icon next to them (on hover) could be solutions, but this might lead to usability or accessibility issues.
> 
> To maintain the current solution where balances are hidden only on the Home screen, I can modify some components, such as `AssetList` and `ActivityList`, to accept an optional `hideBalance` prop. This prop can then be passed down the component tree until it reaches the actual balance component.
> 
> Please let me know what you prefer or if you have any other suggestions!

Thank you for your hard work @adrianocola 💪 

I spoke to the dev team and the design team and we think it's best to keep this simple and manage the state globally with `redux`, your **option 2** above. Our suggestion is to:

- remove the use of a specific context wrapper
- store the privacy setting globally in redux
- toggle privacy mode via the settings menu (get the eye to open / close based on status)

![Screenshot 2024-09-03 at 10 37 00](https://github.com/user-attachments/assets/2ed1af79-4b1d-430f-9465-dccaa8df2abd)

adrianocola commented:
@pete-watters, I forgot about the menu! This is definitively the best solution!

Should I create a new commit for this, or would you prefer I continue squashing them?

pete-watters commented:
> @pete-watters, I forgot about the menu! This is definitively the best solution!
> 
> Should I create a new commit for this, or would you prefer I continue squashing them?
> 
> Edit: I made the change and kept everything in a single commit!

That's great, thank you. One commit is best 👍 

I'll test it soon and hopefully get it merged. Thanks!

pete-watters commented:
> @markmhendrickson @pete-watters we prob need to hide balances here as well? wdyt ![Screenshot 2024-09-04 at 13 16 48](https://private-user-images.githubusercontent.com/46521087/364313148-0428cc7a-a60c-47d9-82ba-fc13778ea486.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU0NDYyNTgsIm5iZiI6MTcyNTQ0NTk1OCwicGF0aCI6Ii80NjUyMTA4Ny8zNjQzMTMxNDgtMDQyOGNjN2EtYTYwYy00N2Q5LTgyYmEtZmMxMzc3OGVhNDg2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTA0VDEwMzIzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRjMGVjMmQ2YmU1OGU5ZDk2YWI2MzZjMTIzMzU5OTJjOWMxOWRhYjdmYWEyZTZiNWZkNWFhZDRhZjhhMmQzZWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.f8ohYHiore0xUsqfuEBTRKIj0JI3Gb4WiZdlyV7N8io) ![Screenshot 2024-09-04 at 13 14 26](https://private-user-images.githubusercontent.com/46521087/364313132-ad5052d7-7393-485c-b342-817eb71276fc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU0NDYyNTgsIm5iZiI6MTcyNTQ0NTk1OCwicGF0aCI6Ii80NjUyMTA4Ny8zNjQzMTMxMzItYWQ1MDUyZDctNzM5My00ODVjLWIzNDItODE3ZWI3MTI3NmZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTA0VDEwMzIzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTllNmY5Zjg3ZjE2MjhjZjU4OGNiYzJiMzI5NzM1MjdlNjU2M2Y0ZjNhN2FlZThkNmRkMzQzZDdiZDU3OGM3OGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.zhkw3jZvmgYDo5VxtMTeH3HGsbboXfOIdMS_d7XaCgs)

I think we could for stacked balances but not in the popup header. In the popup you don't always have access to the settings menu so you won't be able to easily show your balance again. 

It could be nice to hide it everywhere though

pete-watters commented:
Thanks @adrianocola 👍  I tested this and it LGTM. 

I checked with the design team and they asked if you can get rid of the toggle 👁️ on the account card and just have it in the `Settings` menu. 

We also have an open Q here about hiding it in more places https://github.com/leather-io/extension/pull/5830#issuecomment-2328518792. 

It should be good to merge if you could please:
- remove the eye from the `Account` card
- update based on the decision on [hiding more balances](https://github.com/leather-io/extension/pull/5830#issuecomment-2328518792)

Thanks for your patience here and for your contributions

markmhendrickson commented:
Yep good catch

markmhendrickson commented:
> I think we could for stacked balances but not in the popup header. In the popup you don't always have access to the settings menu so you won't be able to easily show your balance again.

It does seem better to hide consistently, even if some places (like the popup header) don't allow toggling per se. The user can open up the extension home and toggle from there as desired, even if not directly in the popup.

adrianocola commented:
Hey, team! Thank you for the input; I'll do it, @pete-watters! 💪

I also like the idea of consistently hiding everywhere. How about allowing the user to click the hidden balance in the header to show the balance? I can even extend this behavior to other places (where the balance doesn't show over a clickable button) to provide an easy way to show balances.


https://github.com/user-attachments/assets/8c34b812-06d6-4722-8ee5-4662056096b9

pete-watters commented:
> Hey, team! Thank you for the input; I'll do it, @pete-watters! 💪
> 
> I also like the idea of consistently hiding everywhere. How about allowing the user to click the hidden balance in the header to show the balance? I can even extend this behavior to other places (where the balance doesn't show over a clickable button) to provide an easy way to show balances.
> 
>  Screen.Recording.2024-09-04.at.08.27.32.mov

Good idea @adrianocola 👍 I checked with the design team and they would be happy if you can go ahead with this. 

Thanks again for your hard work

mica000 commented:
Shall we use *** instead of ••• as the dot seems to get confused with our ellipses?

adrianocola commented:
> Shall we use *** instead of ••• as the dot seems to get confused with our ellipses?

I used dots because asterisks can change drastically depending on the font. An example:
![Screenshot 2024-09-05 at 11 33 07](https://github.com/user-attachments/assets/1394f42c-ec37-470b-9f82-a2a3f5ce29f9)

But I can override and enforce another font, like this example using `Diatype` for the ***:
![Screenshot 2024-09-05 at 11 47 09](https://github.com/user-attachments/assets/7873bcbd-509f-4e02-a987-57ae5436bdc1)

Should I keep the *** using `Diatype` font (or any other font)?

mica000 commented:
> > Shall we use *** instead of ••• as the dot seems to get confused with our ellipses?
> 
> I used dots because asterisks can change drastically depending on the font. An example: ![Screenshot 2024-09-05 at 11 33 07](https://private-user-images.githubusercontent.com/1678020/364824363-1394f42c-ec37-470b-9f82-a2a3f5ce29f9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU2MzE5NTEsIm5iZiI6MTcyNTYzMTY1MSwicGF0aCI6Ii8xNjc4MDIwLzM2NDgyNDM2My0xMzk0ZjQyYy1lYzM3LTQ3MGItOWY4Mi1hMmEzZjVjZTI5ZjkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MDZUMTQwNzMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWE0N2NmMDRlYjA0N2JhZjA1ZWEyMjkyZGMzZGYxNzRiOTM0MWM5ZDRlODQzN2M3N2M3YzZhN2ZmZGM4MzEwNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.PS7IU6Rbwh5RryC5kxEo31p0PPw-Q1M_msVM7t0cJFI)
> 
> But I can override and enforce another font, like this example using `Diatype` for the ***: ![Screenshot 2024-09-05 at 11 47 09](https://private-user-images.githubusercontent.com/1678020/364827988-7873bcbd-509f-4e02-a987-57ae5436bdc1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU2MzE5NTEsIm5iZiI6MTcyNTYzMTY1MSwicGF0aCI6Ii8xNjc4MDIwLzM2NDgyNzk4OC03ODczYmNiZC01MDlmLTRlMDItYTk4Ny01N2FlNTQzNmJkYzEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MDZUMTQwNzMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NWQ5OWE3MjVjYzRiYjdmNWFjMjQ3ZmZiYjVkNjNlZjRlOWFhNjViY2M0Mzg5MWZhMDVjMGUzOTMyNDgwMjRkMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.YIkQ3XpgI7fZdN5Qd43sf5_Ot51vMgHrnOIjU1eRD_I)
> 
> Should I keep the *** using `Diatype` font (or any other font)?

Good idea! Lets keep `Diatype`. Thanks for the suggestion!

adrianocola commented:
Hey @pete-watters, I've made the changes proposed by the team! This is a summary with the changes:

* **Renaming for Consistency:**
  * Refactored terminology from “hide balance” to “privacy mode” for consistent alignment with the interface wording.
  * Renamed component `HideableBalance` to `PrivateText`.
  * Updated Redux store property from `hideBalance` to `isPrivacyMode`.
* **Component Refactoring:**
  * Split the `PrivateText` component into two parts:
    * private-text.layout.tsx (_app/ui/components/privacy_): Stateless presentation component.
    * private-text.tsx (_app/components/privacy_): Handles retrieving privacy mode status from the Redux store and rendering the presentation component.
* **UI Adjustments:**
  * Removed the privacy toggle icon from the account card UI.
  * Added the `PrivateText` component to balances that previously missed this feature.
  * Changed the privacy mask from ••• to ***  and forced the `Diatype` font to maintain consistency across different parts of the extension.
* **New Interaction:**
  * Introduced a feature allowing users to reveal hidden text by clicking the ***. This functionality is enabled for balances in the following sections:
    * Account card on the homepage.
    * Account balance at the bottom of the send page.
    * Account balance at the top of the sign modal.
    
**Let me know if there is anything else I need to change!**

pete-watters commented on src/app/ui/components/privacy/private-text.layout.tsx:
```diff
@@ -0,0 +1,35 @@
+        cursor={canShowValue ? 'pointer' : undefined}
+        display="inline-block"
+        overflow="hidden"
+        fontFamily={isPrivate ? 'Diatype !important' : undefined}
```
Maybe we can use one of [our typography tokens](https://github.com/leather-io/mono/blob/dev/packages/tokens/src/typography.ts) here for the font? Like `heading.01`?

If that's not working that's OK. I think it could be better to force it by using a `style={{` tag instead of `!important`:
```
 style={{ fontFamily:isPrivate ? 'Diatype' : undefined }}
```

pete-watters commented:
> Hey @pete-watters, I've made the changes proposed by the team! This is a summary with the changes:
> 
> * **Renaming for Consistency:**
>   
>   * Refactored terminology from “hide balance” to “privacy mode” for consistent alignment with the interface wording.
>   * Renamed component `HideableBalance` to `PrivateText`.
>   * Updated Redux store property from `hideBalance` to `isPrivacyMode`.
> * **Component Refactoring:**
>   
>   * Split the `PrivateText` component into two parts:
>     
>     * private-text.layout.tsx (_app/ui/components/privacy_): Stateless presentation component.
>     * private-text.tsx (_app/components/privacy_): Handles retrieving privacy mode status from the Redux store and rendering the presentation component.
> * **UI Adjustments:**
>   
>   * Removed the privacy toggle icon from the account card UI.
>   * Added the `PrivateText` component to balances that previously missed this feature.
>   * Changed the privacy mask from ••• to ***  and forced the `Diatype` font to maintain consistency across different parts of the extension.
> * **New Interaction:**
>   
>   * Introduced a feature allowing users to reveal hidden text by clicking the ***. This functionality is enabled for balances in the following sections:
>     
>     * Account card on the homepage.
>     * Account balance at the bottom of the send page.
>     * Account balance at the top of the sign modal.
> 
> **Let me know if there is anything else I need to change!**

Thanks @adrianocola 👏  It's generally looking really great, thanks for all of the hard work, it's almost there.

I mentioned one possible coding style improvement [here](https://github.com/leather-io/extension/pull/5830#discussion_r1751252770)

I noticed that for locked balances the UI is not aligned correctly, so that needs to be fixed: 
![Screenshot 2024-09-10 at 05 58 18](https://github.com/user-attachments/assets/aa552b7d-bcd0-49c1-8311-0a4625c5eda0)

pete-watters commented on src/app/components/balance/balance-private.tsx:
```diff
@@ -0,0 +1,25 @@
+  canClickToShow?: boolean;
+}
+
+export function BalancePrivate({ children, canClickToShow, ...rest }: BalancePrivateProps) {
```
Maybe this can be `PrivateBalance` to match the convention of `PrivateText` ? 

Do you need the prop `canClickToShow`? It seems to be always passed in as `true`?

adrianocola replied:
I renamed the component and am now using it in a place that needs to set the prop `canClickToShow=false` (I actually don't pass it at all).

pete-watters commented on src/app/ui/components/privacy/private-text.layout.tsx:
```diff
@@ -0,0 +1,38 @@
+import React from 'react';
```
Maybe we can introduce a `container` component here - `PrivateText`, so that it can fetch the setting from the state and avoid drilling it down?

So it would be something like `private-text.tsx` 
`
import { PrivateTextLayout } from '@app/ui/components/privacy/private-text.layout'; 

export function PrivateText() {

const isPrivate = useIsPrivateMode();

return (
<PrivateTextLayout
                isPrivate={isBalancePrivate}
                onShowValue={onShowBalance}
                display="inline-block"
                overflow="hidden"
              >
                {balance}
              </PrivateTextLayout>
)
}
```

That could help avoid having to drill down `isPrivate={isPrivate}` so often and we could isolate it into it's own feature: `features/privacy/`. We could also add `onShowValue={onShowBalance}` in there

pete-watters replied:
OK great, thanks for the explanation 👍

pete-watters commented:
> Hey @pete-watters, thanks for the support and patience!
> 
> * Fixed the "locked" alignment issue. Actually, I removed it. It doesn't make much sense to show them anyway while hiding the balance. So, whether there are locked values or not, it will always show only `***`.
> * Using a monospaced font for the asterisks (`fontFamily: 'Fira Code, Consolata, monospace'`). Monospaced fonts center them vertically, making the transition between the balance text and the asterisks more natural.
> * Refactored the code to remove the `PrivateText` stateful component.
>   
>   * This component was being used all over the place to hide balances easily;
>   * But it was also being used inside `*.layout.tsx` components, which broke the project Storybook;
>   * So I'm using the stateless `PrivateTextLayout` component in its place;
>   * I had to get the `isPrivateMode` state from redux everywhere a balance is displayed and pass it to the `PrivateTextLayout` (much like the `isLoading` prop is passed around);

This looks great @adrianocola and it is working perfectly, nice job! 

I added two more suggestions about how we could improve the design of the code but I think it's pretty solid overall and close to merge. Let me know what you think

pete-watters reviewed (approved):
Great work @adrianocola , thanks for your persistence here. The code looks good to me and I tested it and it's working well. 

I will try to get it merged soon but we are having a CI issue now I need to fix first.

@mica000 can you check this build and do a Design QA when you have a chance? 

Here's a demo in case it helps
https://github.com/user-attachments/assets/7d545fd9-b411-4b83-8cde-9f3cad3d02c8

pete-watters commented:
@adrianocola : Great work on this feature 👏  I will get it merged and into the next release once we have resolved the CI issue

mica000 commented:
Indeed! Thanks @adrianocola!! ✨

pete-watters commented:
@adrianocola - I got your work [merged in](https://github.com/leather-io/extension/pull/5865) and we will include it in the next release. Thanks again for your efforts here 👏

adrianocola commented:
Thank you so much, everyone! I’m really glad I could contribute and work with all of you! 🚀
